core/sail/federation/src/main/java/org/eclipse/rdf4j/sail/federation/AbstractFederationConnection.java

```diff
@@ -434,10 +439,27 @@ private TupleExpr optimize(TupleExpr parsed, Dataset dataset, BindingSet binding
 		new QueryMultiJoinOptimizer().optimize(query, dataset, bindings);
 		// new FilterOptimizer().optimize(query, dataset, bindings);
 
-		new EmptyPatternOptimizer(members).optimize(query, dataset, bindings);
+		// prepare bloom filters
+		RepositoryBloomFilter defaultBloomFilter = new AccurateRepositoryBloomFilter(includeInferred);
+		Map<Repository, RepositoryBloomFilter> bloomFilters = federation.getBloomFilters();
+		Map<RepositoryConnection, RepositoryBloomFilter> bloomFiltersByConn = new HashMap<>();
+		List<Repository> repoMembers = federation.getMembers();
+		for(int i=0; i<members.size(); i++) {
```

Comment:
Sorry mark, lost track of this for a bit - but your proposal for an approach with a constructor with a list of pairs sounds good, if that makes things more flexible and robust. If you do this, could you add a comment on issue #215 itself that quickly outlines the changes to the API (and whether or not they are backward compatible)? Just so I have this at hand when the time comes to prep release notes.